destination_middleware.go

```diff
@@ -0,0 +1,236 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdk
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+// DestinationMiddleware wraps a Destination and adds functionality to it.
+type DestinationMiddleware interface {
+	Wrap(Destination) Destination
+}
+
+// DefaultDestinationMiddleware returns a slice of middleware that should be
+// added to all destinations unless there's a good reason not to.
+func DefaultDestinationMiddleware() []DestinationMiddleware {
+	return []DestinationMiddleware{
+		DestinationWithRateLimit{},
+		// DestinationWithBatch{}, // TODO enable batch middleware once batching is implemented
+	}
+}
+
+// DestinationWithMiddleware wraps the destination into the supplied middleware.
+func DestinationWithMiddleware(d Destination, middleware ...DestinationMiddleware) Destination {
+	for _, m := range middleware {
+		d = m.Wrap(d)
+	}
+	return d
+}
+
+// -- DestinationWithBatch -----------------------------------------------------
+
+const (
+	configDestinationBatchSize  = "sdk.batch.size"
+	configDestinationBatchDelay = "sdk.batch.delay"
+)
+
+type ctxKeyBatchEnabled struct{}
+
+// DestinationWithBatch adds support for batching on the destination. It adds
+// two parameters to the destination config:
+//   - `sdk.batch.size` - Maximum size of batch before it gets written to the
+//     destination.
+//   - `sdk.batch.delay` - Maximum delay before an incomplete batch is written
+//     to the destination.
+// To change the defaults of these parameters use the fields of this struct.
+type DestinationWithBatch struct {
+	// DefaultBatchSize is the default value for the batch size.
+	DefaultBatchSize int
+	// DefaultBatchDelay is the default value for the batch delay.
+	DefaultBatchDelay time.Duration
+}
+
+// Wrap a Destination into the batching middleware.
+func (d DestinationWithBatch) Wrap(impl Destination) Destination {
+	return &destinationWithBatch{
+		Destination: impl,
+		defaults:    d,
+	}
+}
+
+// setBatchEnabled stores the boolean in the context. If the context already
+// contains the key it will update the boolean under that key and return the
+// same context, otherwise it will return a new context with the stored value.
+// This is used to signal to destinationPluginAdapter if the Destination is
+// wrapped into DestinationWithBatch middleware.
+func (DestinationWithBatch) setBatchEnabled(ctx context.Context, enabled bool) context.Context {
+	flag, ok := ctx.Value(ctxKeyBatchEnabled{}).(*bool)
```

Comment:
I was wondering why do we use pointers for context values, but it looks it's because we need to be able to update the value, and the context API has no method for that?

Reply:
Right, I'm using a bit of a trick here - `pluginDestinationAdapter` calls this function [here](https://github.com/ConduitIO/conduit-connector-sdk/blob/16670d7a2d469beb2acbcea398d6132afc73a550/destination.go#L89), which stores a pointer in the context. This pointer is then updated in the middleware [here](https://github.com/ConduitIO/conduit-connector-sdk/blob/16670d7a2d469beb2acbcea398d6132afc73a550/destination_middleware.go#L128), this way `pluginDestinationAdapter` can see the updated value. Otherwise we would have to find a different way for signaling back to the adapter that batching is enabled, I don't see a way without changing the `Destination` interface, which I really don't want to do.